WinObjC Bolts sample

venkatkongara opened the pull request:
First Bolts sample app for utilizing Tasks from the API

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/microsoft/winobjc-samples/12)
<!-- Reviewable:end -->

Commit:
WinObjC Bolts sample

venkatkongara commented:
@antoine-haas is added to the review. #Closed

venkatkongara commented:
@asimonov-msft is added to the review. #Closed

venkatkongara commented:
@andrejo-msft is added to the review. #Closed

venkatkongara commented:
@JeffWells39654 is added to the review. #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,649 @@
+			isa = PBXProject;
+			attributes = {
+				LastUpgradeCheck = 0820;
+				ORGANIZATIONNAME = "venkat kongaraSiva";
```
Personally identifiable information. Switch to Microsoft. #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/AppDelegate.h:
```diff
@@ -0,0 +1,17 @@
+//  BoltsInClasspath
+//
+//  Created by siva kongara on 1/14/17.
+//  Copyright © 2017 venkat kongaraSiva. All rights reserved.
```
>2017 venkat kongaraSiva [](start = 16, length = 23)

Venkat, you'll want to use the same Copyright header as the rest of the code in -samples, updated for 2017

(sample)
[https://github.com/Microsoft/WinObjC-Samples/blob/master/Scenarios/Maps/Maps/MapsSample/AppDelegate.h](https://github.com/Microsoft/WinObjC-Samples/blob/master/Scenarios/Maps/Maps/MapsSample/AppDelegate.h)

//******************************************************************************
//
// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
//
// This code is licensed under the MIT License (MIT).
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
//******************************************************************************


 #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/AppDelegate.m:
```diff
@@ -0,0 +1,50 @@
+//
+//  Created by siva kongara on 1/14/17.
+//  Copyright © 2017 venkat kongaraSiva. All rights reserved.
+//
```
same #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/Constants.h:
```diff
@@ -0,0 +1,14 @@
+//  Constants.h
+//  BoltsInClasspath
+//
+//  Created by siva kongara on 1/18/17.
```
same #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/Constants.m:
```diff
@@ -0,0 +1,13 @@
+//  Constants.m
+//  BoltsInClasspath
+//
+//  Created by siva kongara on 1/18/17.
```
same #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/main.m:
```diff
@@ -0,0 +1,16 @@
+//  BoltsInClasspath
+//
+//  Created by siva kongara on 1/14/17.
+//  Copyright © 2017 venkat kongaraSiva. All rights reserved.
```
same #Closed

andrejo-msft replied:
Still missing the MIT license.

---
In reply to: [97418217](https://github.com/Microsoft/WinObjC-Samples/pull/12#discussion_r97418217) [](ancestors = 97418217)

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/TaskExecutionResource.h:
```diff
@@ -0,0 +1,20 @@
+//
+//  TaskExecutionResource.h
+//  WinObjcBoltsTest
```
same #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/TaskExecutionResource.m:
```diff
@@ -0,0 +1,36 @@
+//
+//  TaskExecutionResource.m
+//  WinObjcBoltsTest
```
...and so on. We need the Microsoft MIT License header in your BoltsInClasspath/BoltsInClasspath files to push this to develop. #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+The sample uses weather.com free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from Weather.com and replace it 
+with apiKey placeholder in the Constants.m file
```
Add the link to weather.com site where you get the token. #WontFix

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+The sample uses weather.com free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from Weather.com and replace it 
+with apiKey placeholder in the Constants.m file
+
+Note: Make sure this folder and Winobjc release folders exist in the same directory. otherwise it may not run.
```
What does this mean? Why does this folder and the release folder need to be in the same directory? #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+
+Note: Make sure this folder and Winobjc release folders exist in the same directory. otherwise it may not run.
+
+In the BoltsInClassPath sample app to make this app run on WINOBJC I have found two issues in the app build phase.
```
s/WinObjC/Windows platforms #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+ Bolts.lib (BFExecutor_F456FB2F.obj) : error LNK 2019: unresolved external symbol _pthread_get_stackaddr_np referenced in function remaining_stack_size
+ Bolts.lib (BFExecutor_F456FB2F.obj) : error LNK 2019: unresolved external symbol _pthread_get_stacksize_np referenced in function remaining_stack_size
+
+This happens because I see a different implementations for pThread on Winobjc. so my temp workaround is
```
same #Closed

andrejo-msft commented:
Venkat, I see that you are adding the entire Bolts framework.

Rather than maintain the present Bolts framework for your app, I want to recommend that you use git submodule instead. 

Check out the README.md for WinObjC tools as an example. ([https://github.com/Microsoft/WinObjC/tree/develop/tools](https://github.com/Microsoft/WinObjC/tree/develop/tools)). I'll update your readme.MD with a suggested way to document this.

 #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+This happens because I see a different implementations for pThread on Winobjc. so my temp workaround is
+
+go to your Bolts dependency in the solution and navigate to Bolts/Bolts/Common/BFExecutor.m file. change the below method implementation like this.
+
```
As I mentioned in the Discussion, I recommend the use of git submodule instead of dragging along the complete framework.

In this case, I suggest that you provide directions to add the submodule, then update the Bolts framework as you suggest below. #Closed

andrejo-msft replied:
The code below should use bolding to show which line should be added/subtracted.

---
In reply to: [97421875](https://github.com/Microsoft/WinObjC-Samples/pull/12#discussion_r97421875) [](ancestors = 97421875)

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+
+
+2) go to your Bolts dependency in the solution and navigate to Bolts/Public Headers/Bolts.h file. change the line in import statements like this.
+	#if __has_include(<Bolts/BFAppLink.h>) && WINOBJC && !TARGET_OS_WATCH && !TARGET_OS_TV
```
As above.  #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md.txt:
```diff
@@ -0,0 +1,73 @@
+
+Note: APPLinks part of BoltsAPI was not yet implemented.
+
+All the methods listed below are used in sample and are tested on both iOS and Windows environment.
```
s/tested/tested successfully #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspath/ViewController.m:
```diff
@@ -0,0 +1,275 @@
+    [[self view] setBackgroundColor:[UIColor colorWithWhite:0.5 alpha:1.0]];
+    
+    UILabel* label1 = [[UILabel alloc] init];
+    [label1 setText:@"Bolts Sample Application.\nPlease observe console for output."];
```
This works well for desktop, but not for mobile device. Consider an overall 'pass' or 'fail' next to the buttons.

I would like to recommend one more trick. Run each positive test when the app starts up as part of constructView. This allows a developer to pick it up, build it, run it, and now that everything is working.  #WontFix

andrejo-msft commented:
:shipit:

andrejo-msft commented:
Venkat, it's a good start. Overall problems:
- Privacy and compliance issues have to be completed before we move forward.
- Providing the entire Bolts framework is too much. I recommend using git submodule to pull down the appropriate code, then changing it.
 #Closed

Commit:
added Bolts framework as a submodule and deleted the framework files from the BoltsInClassPath sample app.

Commit:
changed copyrights and modified README.md file

Commit:
included bolts submodule hash

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/BoltsInClasspathTests/BoltsInClasspathTests.m:
```diff
@@ -0,0 +1,39 @@
+//  BoltsInClasspathTests.m
+//  BoltsInClasspathTests
+//
+//  Created by siva kongara on 1/14/17.
```
Still missing the MIT header. #Closed

andrejo-msft commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "ThirdParty/Bolts/BoltsInClasspath/Bolts-ObjC-1.8.4"]
+	path = ThirdParty/Bolts/BoltsInClasspath/Bolts-ObjC-1.8.4
+	url = https://github.com/BoltsFramework/Bolts-ObjC.git
```
This is stored in the database itself. I'm assuming the hash is for Bolts 1.8.4. #Closed

venkatkongara replied:
yes #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,78 @@
+Make sure you are using the Bolts 1.8.4 version submodule which has hash
+	e64deecb2f0e10ac0dbb71f522c7a5b9cafb0b4d
+	
+Then follow the steps from WinObjC github repo documentation to create a visual studio solution.
```
Add the link to weather.com so that the user does not need to search for it. #Closed

andrejo-msft commented:
Venkat, just a few issues left. Once you take care of them, we'll ship it.
- Add the MIT License header where I point it out.
- Read the remaining comments and address them.

The change I mention to ViewController could be done as part of the next App Links work. Please let me know what you think of running the app's actions from the start instead of interactively.
 #Closed

venkatkongara committed:
Update main.m

MIT license

venkatkongara committed:
Update BoltsInClasspathTests.m

venkatkongara committed:
Update README.md

included https://home.openweathermap.org/users/sign_up site address to let the user register.

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,77 @@
+﻿The sample uses weather.com free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from https://home.openweathermap.org/users/sign_up and replace it with apiKey placeholder in the Constants.m file
+
+This project uses bolts as a dependency.When you first clone the repo if you have included ‘—recursive’ skip the next statement. Else
+	Navigate to samples root directory and run the ‘git submodule update --init --recursive’ git command.
```
>git submodule update --init --recursive [](start = 49, length = 39)

nit: substitute 'samples' for 'WinObjC-samples' root directory #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,77 @@
+	#if __has_include(<Bolts/BFAppLink.h>) && WINOBJC && !TARGET_OS_WATCH && !TARGET_OS_TV
+
+
+Now Build and run the sample on to simulator.
```
I followed the directions above and got to a situation where I could not successfully build the app.

My process:
- clone your repo from venkat-kongara/winobjc-samples (develop) on my Windows machine
- run git submodule update --init --recursive from the winobjc-samples root directory.

The repo I saw was not the 1.8.4 release hash. Instead, it was the most recent master hash. I reset it to the 1.8.4 hash, made the suggested changes and built. I saw a significant number of build errors.

We need to understand why your build is working and mine is not, then update the documentation so that anyone pulling down your code can replicate it. #Closed

andrejo-msft commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,77 @@
+}
+
+
+2) go to your Bolts dependency in the solution and navigate to Bolts/Public Headers/Bolts.h file. change the line in import statements like this.
```
Add the enable ARC step (and file an issue on this in Github). #Closed

antoine-haas commented on ThirdParty/README.md:
```diff
@@ -0,0 +1 @@
+TBD
```
This should read:
    # Third Party
    Sample apps demonstrating common third party frameworks scenarios. #Closed

venkatkongara committed:
Update README.md

venkatkongara committed:
Update README.md

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,77 @@
+﻿The sample uses weather.com free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from https://home.openweathermap.org/users/sign_up and replace it with apiKey placeholder in the Constants.m file
```
The readme should have at least a title and use GitHub markdown to make it more readable by having important text in bold and Objective-C code properly highlighted. See https://github.com/Microsoft/WinObjC-Samples/tree/master/Scenarios/Radial%20Control or any other readme from the scenario folder for good examples 

venkatkongara committed:
Update README.md

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+﻿The sample uses openweathermap.org free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from https://home.openweathermap.org/users/sign_up and replace it with apiKey placeholder in the Constants.m file
+
+This project uses bolts as a dependency.When you first clone the repo if you have included ‘--recursive’ skip the next statement. Else
```
Space missing after "dependency."

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+﻿The sample uses openweathermap.org free subscription rest API to demonstrate the Bolts usage. Please register for free and use the token generated from https://home.openweathermap.org/users/sign_up and replace it with apiKey placeholder in the Constants.m file
+
+This project uses bolts as a dependency.When you first clone the repo if you have included ‘--recursive’ skip the next statement. Else
+	Navigate to samples root directory and run the ‘git submodule update --init --recursive’ git command.
```
"Navigate" should have lowercase n

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+	e64deecb2f0e10ac0dbb71f522c7a5b9cafb0b4d
+	(you can do that by navigating to the WinObjC-Samples/ThirdParty/Bolts/BoltsInClasspath directory in cmd and type in git checkout 1.8.4)
+	
+Then follow the steps from WinObjC github repo documentation to create a visual studio solution.
```
"WinObjC GitHub repo documentation" should link to https://github.com/Microsoft/WinObjC/#getting-started-with-the-bridge #Closed

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+
+In Visual Studio expand the solution and navigate to Bolts folder. Then right click on Bolts target go to properties->clang->Enable ObjectiveC ARC and set it to YES(-fobjc-arc).Do this for the other target BoltsInClassPath also in subsequent BoltsInClassPath folder.
+
+In the BoltsInClassPath sample app to make this app run on Windows I have found two issues in the app build phase.
```
You should replace "I" in the readme by "you", so this sentences read "To make the BoltsInClassPath sample run on Windows, you will have to work around two issues :" for example. This readme is for our users on GitHub to read and follow as a tutorial and should be addressed to them. #Closed

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+
+In the BoltsInClassPath sample app to make this app run on Windows I have found two issues in the app build phase.
+
+1) When you start building app you may encounter
```
"the" missing. Should read "start building the app " #Closed

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+ Bolts.lib (BFExecutor_F456FB2F.obj) : error LNK 2019: unresolved external symbol _pthread_get_stackaddr_np referenced in function remaining_stack_size
+ Bolts.lib (BFExecutor_F456FB2F.obj) : error LNK 2019: unresolved external symbol _pthread_get_stacksize_np referenced in function remaining_stack_size
+
+This happens because I see a different implementations for pThread on Winobjc. so my temp workaround is
```
This should also be rephrased to address external users: "This happens because WinObjC uses a different implementation for pThread. On Windows, you'll have to go to your Bolts dependency in the solution and navigate to the Bolts/Bolts/Common/BFExecutor.m file. Change the below method implementation as shown:" #Closed

andrejo-msft replied:
Please remove the quotes.

---
In reply to: [98303358](https://github.com/Microsoft/WinObjC-Samples/pull/12#discussion_r98303358) [](ancestors = 98303358)

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+	#if __has_include(<Bolts/BFAppLink.h>) && WINOBJC && !TARGET_OS_WATCH && !TARGET_OS_TV
+
+
+Now Build and run the sample on to simulator.
```
This should read: "Now build and run the sample on the emulator". Visual Studio uses the term "emulator" instead of "simulator" :) #Closed

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+
+Now Make sure you are using the Bolts 1.8.4 version submodule which has hash
+	e64deecb2f0e10ac0dbb71f522c7a5b9cafb0b4d
+	(you can do that by navigating to the WinObjC-Samples/ThirdParty/Bolts/BoltsInClasspath directory in cmd and type in git checkout 1.8.4)
```
the path you should navigate to is: winobjcSamples\WinObjC-Samples\ThirdParty\Bolts\BoltsInClasspath\Bolts-ObjC-1.8.4 #Closed

antoine-haas commented on ThirdParty/Bolts/BoltsInClasspath/README.md:
```diff
@@ -0,0 +1,81 @@
+
+
+2) go to your Bolts dependency in the solution and navigate to Bolts/Public Headers/Bolts.h file. change the line in import statements like this.
+	#if __has_include(<Bolts/BFAppLink.h>) && WINOBJC && !TARGET_OS_WATCH && !TARGET_OS_TV
```
this is being displays as #if __has_include() && WINOBJC && !TARGET_OS_WATCH && !TARGET_OS_TV by GitHub markdown, so the Bolts/BFAppLink.h> is missing. Putting this between GitHub markdown tags for code ​
```Objective-C ... ``` should fix it. #Closed

venkatkongara committed:
Update README.md

venkatkongara committed:
Update README.md

venkatkongara committed:
Update README.md

venkatkongara committed:
Update README.md

Commit:
explicitly made the execution happen on main thread to fix errors.

Commit:
changed the language grammer in README.md

andrejo-msft commented:
Please complete the last set of comments so we can close this CR.

venkatkongara committed:
Update README.md

venkatkongara committed:
Update Constants.m

venkatkongara committed:
Update README.md

venkatkongara committed:
Update README.md
